(qa-3114) update Rakefile to rototiller 1.y

er0ck opened the pull request:
(QA-3114) update our Rakefile to rototiller 1.0+

    (maint) add test namespace to Rakefile, clarify descriptions of tasks

    * remove descriptions of tasks that aren't useful to users so they don't
    show up in -T
    * add test namespace, rename spec task to test:unit
    * add acceptance task description to remove rototiller default

    (maint) describe rake tasks with -T as default task

Commit:
(maint) describe rake tasks with -T as default task

Commit:
(maint) add test namespace to Rakefile, clarify descriptions of tasks

* remove descriptions of tasks that aren't useful to users so they don't
show up in -T
* add test namespace, rename spec task to test:unit
* add acceptance task description to remove rototiller default

Commit:
(QA-3114) update our Rakefile to rototiller 1.0+

er0ck commented:
this probably needs updating to enable migration to ci.next.

@smcelmurry, thoughts?  i forget what tasks we need to enable RMM/ABS

junebug-spencer commented:
@er0ck Yeah the bhg string will need an optional flag for the abs hypervisor (`--hypervisor abs`). You can probably copy what is being done over in [pe_acceptance_tests](https://github.com/puppetlabs/pe_acceptance_tests/blob/2017.3.x/acceptance/high_availability/rakefile.rb#L29).

You'll also need to to add beaker-abs to the Gemfile:
`gem "beaker-abs", *location_for(ENV['BEAKER_ABS_VERSION'] || "~> 0.2")`

Thanks for checking in with this!

Commit:
(maint) add beaker-abs to Gemfile to enable move to ci.next

er0ck commented:
thanks @smcelmurry.

in the pe_acceptance_tests HA setup they use `BEAKER_ABS`
is that typical?  i can't see any jobs that use this variable.
what is the typical value?

i'm realizing rototiller has trouble with 'boolean' env vars, especially when it's a default false/empty

Commit:
(maint) add BEAKER_ABS env var and bhg option

* if BEAKER_ABS is set, to anything, use --hypervisor abs in BHG
* rototiller currently doesn't have a good story around this use case
  * so we use a work around where we add messaging to a global env_var,
   but set the option and its value implicitly in the command using an
   argument. :-(
* we also add messaging to LAYOUT and TEST_TARGET
  * and move LAYOUT env_var into bhg task, out of the acceptance task

samwoods1 commented:
@er0ck BEAKER_ABS was the ENV var that @sbeaulie suggested and said he would ensure would be there in the ci-next job once HA got migrated.

er0ck commented:
thanks @samwoods1 note the way we're using it in the Rakefile in this PR.  it's not a great use case for rototiller.  i think i've worked around it ok.  it's just nasty because of limitations in rototiller.  
if we think having default-false boolean env vars like this is the right interface, we'll just have to fix it in future rototiller to make our syntax more clear.

samwoods1 commented on Rakefile:
```diff
@@ -2,82 +2,90 @@ require 'rspec/core/rake_task'
+      bhg_command.add_argument do |arg|
+        arg.name = 'centos7-64'
+        arg.add_env({:name => 'LAYOUT', :message => 'The beaker-hostgenerator pattern'})
+        arg.add_env({:name => 'TEST_TARGET', :message => 'The beaker-hostgenerator pattern'})
```
What is the difference between LAYOUT and TEST_TARGET? Should they have unique descriptions?

er0ck replied:
no difference, LAYOUT is historical, but not very descriptive.  the PA components have moved to TEST_TARGET, or similar

er0ck replied:
i suppose we could state that TEST_TARGET value is used over LAYOUT, if they are both present.  (rototiller defaults to `last`)

samwoods1 commented on Rakefile:
```diff
@@ -2,82 +2,90 @@ require 'rspec/core/rake_task'
+      bhg_command.name = "beaker-hostgenerator"
+      bhg_command.add_argument do |argument|
+        # ugh, this is terrible, but it's the only way in current rototiller
+        argument.name = ENV['BEAKER_ABS'] ? ' --hypervisor=abs' : ''
```
Yeah... if the env var is a string is't not bad, but if it's a true/false value that's translating into a string or no string, this seems like the best way given current implementation.

samwoods1 commented:
👍

er0ck commented:
@puppetlabs-jenkins retest me please
just to make ci green because [skip ci] doesn't work on scratch PRs??

er0ck commented:
@puppetlabs-jenkins retest this please

er0ck commented:
@puppetlabs-jenkins retest this please

er0ck commented:
if we go with this task name, we have to update the `command` in ci-job-configs

er0ck commented:
#82 is partly superseded by this.  

I'll fix this one up, get it in, then we can rebase 82

er0ck commented:
@samwoods1 @zreichert can one of you push go on this, if it looks good?  i've added backwards compat task names that surprisingly work (you can have a task named the same thing as a namespace 😢 ).  

This is gating 82 and the move to ci.next, gating addition of periodic runs, move to travis, etc

edit: sam gave the 👍 on this.  i'll give it a day and then hit go if no one else does
